crates/oxc_linter/src/rules/typescript/no_restricted_types.rs

```diff
@@ -0,0 +1,557 @@
+/// Can be:
+/// - `true` - ban with default message
+/// - A string - ban with custom message
+/// - An object with `message` and optional `fixWith` and `suggest`
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
+#[serde(untagged)]
+enum BanConfigValue {
+    /// `"TypeName": true` - ban with default message
+    /// Note: Only `true` is valid; `false` would fail to deserialize and be ignored.
+    Bool(True),
+    /// `"TypeName": "message"` - ban with custom message
+    Message(String),
+    /// `"TypeName": { "message": "...", "fixWith": "...", "suggest": ["..."] }` - full config
+    Object {
+        #[serde(default)]
+        message: Option<String>,
+        #[serde(default, rename = "fixWith")]
+        fix_with: Option<String>,
+        #[serde(default)]
+        suggest: Option<Vec<String>>,
+    },
+}
+
+/// A type that only deserializes from `true`.
+/// This matches the upstream typescript-eslint schema which only allows `true`, not `false`.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, JsonSchema)]
+struct True;
+
+impl<'de> Deserialize<'de> for True {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let value = bool::deserialize(deserializer)?;
+        if value { Ok(True) } else { Err(serde::de::Error::custom("expected `true`, got `false`")) }
+    }
+}
+
+impl BanConfigValue {
+    fn message(&self) -> Option<&str> {
+        match self {
+            BanConfigValue::Bool(_) => None,
+            BanConfigValue::Message(msg) => Some(msg.as_str()),
+            BanConfigValue::Object { message, .. } => message.as_deref(),
+        }
+    }
+
+    fn fix_with(&self) -> Option<&str> {
+        match self {
+            BanConfigValue::Bool(_) | BanConfigValue::Message(_) => None,
+            BanConfigValue::Object { fix_with, .. } => fix_with.as_deref(),
+        }
+    }
+
+    fn suggest(&self) -> Option<&[String]> {
+        match self {
+            BanConfigValue::Bool(_) | BanConfigValue::Message(_) => None,
+            BanConfigValue::Object { suggest, .. } => suggest.as_deref(),
+        }
+    }
+}
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Disallow certain types from being used.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Some built-in types have aliases, while some types are considered dangerous or harmful.
+    /// It's often a good idea to ban certain types to help with consistency and safety.
+    ///
+    /// ### Examples
+    ///
+    /// Given `{ "types": { "Foo": { "message": "Use Bar instead", "fixWith": "Bar" } } }`:
+    ///
+    /// Examples of **incorrect** code for this rule:
+    /// ```ts
+    /// let value: Foo;
+    /// ```
+    ///
+    /// Examples of **correct** code for this rule:
+    /// ```ts
+    /// let value: Bar;
+    /// ```
+    NoRestrictedTypes,
+    typescript,
+    restriction,
+    fix_suggestion,
```

Comment:
Should fix the test failure:
```suggestion
    fix_suggestion,
    config = NoRestrictedTypesConfig,
```

Reply:
lol thanks. i thought i did this, clearly not